### PR TITLE
Refactored benchmark reporting mechanism.

### PR DESCRIPTION
**Pull request by dgquintas**
It now allows pluggging in "reporter" instances to process the benchmark results arbitrarily.
This would allow, for example, to send results to a leaderboard and/or other systems for tracking performance metrics.

### CODE REVIEW

**Comment by ctiller** [comment]
What's the advantage of a registry versus a simple switch statement?

**Comment by dgquintas** [comment]
Right now there's only one type of reporter, the logging one. We need to
add more (to forward the results to performance tracking systems).
Moreover, some reporters will be available only in some circumstances
(internal/external code). By using a properly configured registry and
passing it to the actual benchmarks (which I've realized is a change I need
to introduce in the pull request, that's not currently the case), we can
modify behavior without interfering with any of the individual benchmarks'
code, only with the main(), of which we can have different versions
depending on the context (again, internal/external).

What would the switch discriminate over? There can be an arbitrary number
of reporters, all of them receiving the results info. Different benchmarks
are also interested in different types of reports (QPS+Latency, only
latency, etc.)

**Comment by ctiller** [comment]
If we have different main functions then why not just pass a reporter
instance in?

My experience says that most registries need not exist, and that there's
usually a simpler, more direct mechanism available. I'm looking to find
what that is in this case.

**Comment by dgquintas** [comment]
The naming is unfortunate. It's simply a container for a collection of
Reporter instances. Indeed in the main we can simply provide a set of these
instances to the function running the benchmark and then iterate calling
the Report() method, which is just what the so-called registry does now.

So yeah, we can do without the registry class itself and its Register
method, simply have the benchmark methods receive a collection of Reporter
instances to consider. If that sounds good I'll make the changes in a
jiffy, going back home now.

**Comment by dgquintas** [comment]
In the midst of some changes based on comments and to allow for internal/external config for benchmarks (a-la tests).

**Comment by dgquintas** [comment]
Done. PTAL.

File: test/cpp/qps/qps_test.cc [hunk 1]
```diff
@@ -64,16 +67,23 @@ static void RunQPS() {
   const auto result =
       RunScenario(client_config, 1, server_config, 1, WARMUP, BENCHMARK, -2);
 
-  ReportQPSPerCore(result, server_config);
-  ReportLatency(result);
+  std::set<ReportType> types;
+  types.insert(grpc::testing::ReportType::REPORT_QPS_PER_CORE);
+  types.insert(grpc::testing::ReportType::REPORT_LATENCY);
+  for (const auto& reporter : reporters) {
+    reporter->Report({client_config, server_config, result}, types);
```
Comment (vjpai) [inline test/cpp/qps/qps_test.cc]: Sorry, some old compilers that we support don't like this kind of brace initialization.

**Comment by vjpai** [comment]
Please ignore my previous deleted comment about old-compiler compatibility.

**Comment by vjpai** [comment]
For starters, let me say that I believe that having a refactored reporting mechanism is the right way to do this. I'm not going to nitpick right now, but there are a lot of line notes that need to get added here (range-based fors and all the other renegades out there).

My concern is that I don't believe that this is the right way to do this. I'd personally greatly prefer to see something based on having a vector of reporting functors that can be naturally extended by the user. This is the way, IMO, to extend this naturally to support formats for external reporting tools that we have no control over. The existing reporting functions like ReportQPS should continue to exist as samples and builtins for convenience, but the current enum/switch/case structure basically makes them the be-all and end-all. I don't think that that's the way to go.

**Comment by dgquintas** [comment]
Grrr, I keep forgetting the old compilers. Sorry about that. On the plus
side, it surfaced the problem with Travis.

Are you proposing simply having a collection or arbitrary functors that can
do anything they want? That's already the case, except that some structure
is given in terms of what's being reported. I can see the use in removing
these constraints and just have the functors do whatever, even things we
haven't foreseen (outside of QPSs, latencies and times). Is this your
proposal?

**Comment by ctiller** [comment]
So the problem space (as I see it) is thus:

we have tests for which some set of statistics makes sense/are calculable:
{a, b, c} (qps and cpu usage on a ping pong test are mostly noise)
we have use cases that would like to get those statistics to some place:
{stdout, leader-board}
we'd like to have binaries with main() functions that can stay untouched as
each of these sets mutates over time

Can I suggest that report.cc contain some gflag arguments that control what
to include (probably two flags)
And provide a class, QpsReport? structured like so:

class QpsReport {
 public:
  // call these functions to mark some set of stats as 'sensible' for a
benchmark - mirrors existing functions
  QpsReport& ProvidesQPS();
  QpsReport& ProvidesQPSPerCore();

  // finally, call this method
  void Report(const ScenerioResult& result, const ClientConfig& cli_config,
const ServerConfig& svr_config);
};

Individual tests can do:
QpsReport().ProvidesQPS().Report(...);

Report can gather statistics and farm out work as it deems fit, using its
flags to either filter the stats, or decide where to push results to.

**Comment by dgquintas** [comment]
This problem is trickier than it seems. Having {Qps,Latency,Time}Report is
nice, but the Report() method only kicks the ball forward: firstly, it
needs to know to which reporters (possibly more than one) to push the data
to. How it knows to which ones could indeed be done via flags, however that
introduces a changes in behavior driven by a global flag, not visible in
the interface. Regardless, the issue then is that all we have at this point
is knowledge of the type of report we want (qps, qps/core, latency, time).
If we somehow know we want to push to reporters A and B, which can be
simple functions, we need to pass both the type of report on top of all the
data, because at this point we don't know how A or B will need to format
the data (they may need to populate protos, etc.). In the end we have the
same (all-data + type) interface as in the current pull req, only further
down the call stack. Both A and B will need a switch inside discriminating
on the metric type. In addition, we need the flexibility to have
internal/external reporters, meaning the reporters have to be quite
decoupled from everything else.

In sum, because the user needs the ability to 1) choose from a set of
metrics to report and 2) choose to which systems to push them too, I still
think the current approach is appropriate. Choosing the metric is done by
Report.Metric() as opposed to Metric.Report(). Choosing to which systems to
push to is configured in benchmark_config.cc (based on flags introduced and
reduced only to that same file -this needs to be added to the pull. req-).
This also covers the internal/external flavors of the file, same as
test_config currently does.

If we wanted to have arbitrary functors, we'd need to pass the type of
metric we want. This is the same as having a method named after the metric,
only more fragile. Adding a new metric requires code changes in both cases,
but without the compiler's help for the functors case.

Currently, a random user wanting to implement a new reporter for his
home-grown leaderboard would only need to implement the Reporter interface,
with its four virtual methods for the currently supported metrics. If new
metrics were to be introduced, because the virtual methods aren't pure,
nothing would break. The user, at her leisure, could implement support for
the new metrics.
In addition, the single file that'd need to be modified to add this new
reporter is benchmark_config, possibly to add a new controlling flag that's
checked in the InitBenchmarkReporters() to add to the collection of
reporters the benchmark sends its results to.

**Comment by dgquintas** [comment]
Ok, the saga continues. The code has been simplified, incorporating Craig's comment in part, which resulted in a lot of spuriously introduced new code being deleted.

PTAL and thanks for all the comments, goes without saying :)

**Comment by ctiller** [comment]
LGTM. The only additional comment I have is pure bikeshedding, but I'm going to make it anyway :)

If you introduced a CompositeReporter (in the obvious way), and changed InitBenchmarkReporters() to shared_ptr<Reporter> GetReporter() you could reduce the amount of infrastructure you need to carry in the driver code a little more:

```c
for (const auto& reporter : reporters) {
  reporter->ReportQPS(result);
  reporter->ReportLatency(result);
}
```

would become:

```c
GetReporter()->ReportQPS(result);
GetReporter()->ReportLatency(result);
```

... and with a little more refactoring, providing some global wrappers around GetReporter->... we'd get back to exactly the code that's in the drivers today.

The motivator for the above is that I'd like to keep the various drivers as 'script-like' as possible, and keep the infrastructure to drive them as separated as possible, as I think this gives us a really good story for how we'll run and use the driver programs.

**Comment by dgquintas** [comment]
Duh! Great call @ctiller, this is almost the textbook example of the composite pattern.

PTAL.

**Comment by ctiller** [comment]
Prefer return by value for GetReporter and LGTM.

**Comment by dgquintas** [comment]
About the return of the shared_ptr, I'm curious why you'd prefer to return by value. All methods are const, it seems that returning by value would unnecessarily increment the reference count. 
On the other hand, this is a minute optimization that is also dependent on the current const semantics of the methods. What's your take?

**Comment by dgquintas** [comment]
To answer my own question: in anything more complex than this very simple setup, there's the risk the pointer will be deallocated while still refs to it are laying around.

**Comment by ctiller** [comment]
Return by value makes refactoring later much easier (you're not obligated
to have an instance around), so unless there's a compelling reason not to
do so I tend to prefer it.

**Comment by dgquintas** [comment]
Agreed.

Updated. PTAL.

**Comment by dgquintas** [comment]
Greenness!

**Comment by dgquintas** [comment]
@vjpai could you merge when you get a chance if it LGTY?

File: test/cpp/util/benchmark_config.h [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+#include "test/cpp/qps/report.h"
+
+namespace grpc {
+namespace testing {
+
+void InitBenchmark(int* argc, char*** argv, bool remove_flags);
+
+/** Returns the benchmark Reporter instance.
+ *
+ * The returned instane will take care of generating reports for all the actual
```
Comment (vjpai) [inline test/cpp/util/benchmark_config.h]: instane -> instance

**Comment by vjpai** [comment]
LGTM, please fix the typo.

**Comment by dgquintas** [comment]
C# tests are still problematic, but now that we are post-launch, I think it'd be safe to merge.

**Comment by dgquintas** [comment]
OMG GREEN!!!1one

**Pull request merged**